tests/engine/processes/workchains/test_restart.py

```diff
@@ -8,44 +8,94 @@
+@pytest.mark.usefixtures('clear_database_before_test')
+def test_excepted_process(generate_work_chain, generate_calculation_node):
+    """Test that the workchain aborts if the sub process was excepted."""
+    process = generate_work_chain(SomeWorkChain, {})
+    process.setup()
+    process.ctx.children = [generate_calculation_node(ProcessState.EXCEPTED)]
+    assert process.inspect_process() == BaseRestartWorkChain.exit_codes.ERROR_SUB_PROCESS_EXCEPTED  # pylint: disable=no-member
+
+
+@pytest.mark.usefixtures('clear_database_before_test')
+def test_killed_process(generate_work_chain, generate_calculation_node):
+    """Test that the workchain aborts if the sub process was killed."""
+    process = generate_work_chain(SomeWorkChain, {})
+    process.setup()
+    process.ctx.children = [generate_calculation_node(ProcessState.KILLED)]
+    assert process.inspect_process() == BaseRestartWorkChain.exit_codes.ERROR_SUB_PROCESS_KILLED  # pylint: disable=no-member
+
+
+@pytest.mark.usefixtures('clear_database_before_test')
+def test_unhandled_failure(generate_work_chain, generate_calculation_node):
+    """Test the unhandled failure mechanism.
+
+    The workchain should be aborted if there are two consecutive failed sub processes that went unhandled. We simulate
+    it by setting `ctx.unhandled_failure` to True and append two failed process nodes in `ctx.children`.
+    """
+    process = generate_work_chain(SomeWorkChain, {})
+    process.setup()
+    process.ctx.unhandled_failure = True
+    process.ctx.iteration = 2
+    process.ctx.children = [generate_calculation_node(exit_status=100) for _ in range(2)]
+    assert process.inspect_process() == BaseRestartWorkChain.exit_codes.ERROR_SECOND_CONSECUTIVE_UNHANDLED_FAILURE  # pylint: disable=no-member
+
+
+@pytest.mark.usefixtures('clear_database_before_test')
+def test_unhandled_reset_after_success(generate_work_chain, generate_calculation_node):
+    """Test `ctx.unhandled_failure` is reset to `False` in `inspect_process` after a successful process."""
+    process = generate_work_chain(SomeWorkChain, {})
+    process.setup()
+    process.ctx.unhandled_failure = True
+    process.ctx.iteration = 2
+    process.ctx.children = [generate_calculation_node(exit_status=100), generate_calculation_node(exit_status=0)]
+    assert process.inspect_process() is None
+    assert process.ctx.unhandled_failure is False
+
+
+@pytest.mark.usefixtures('clear_database_before_test')
+def test_unhandled_reset_after_handled(generate_work_chain, generate_calculation_node):
+    """Test `ctx.unhandled_failure` is reset to `False` in `inspect_process` after a handled failed process."""
+    process = generate_work_chain(SomeWorkChain, {})
+    process.setup()
+    process.ctx.unhandled_failure = True
+    process.ctx.iteration = 2
+    process.ctx.children = [generate_calculation_node(exit_status=0), generate_calculation_node(exit_status=400)]
```

Comment:
If you are just testing that the handler catches the `exit_status=400` here, do you need the previous calculation with `exit_status=0`?

Reply:
Technically no, since the code currently won't check the first, and I don't run `inspect_process` manually twice, after having appended a calculation to the ctx first each time. I will maybe change the test to do this and be more explicit, instead of setting the `iteration` manually